### DISCUSSION TITLE
Transient

### DISCUSSION THREAD

**Discussion by semi710** [marlonrichert/zsh-snap#265] [Q&A]
Hey!,
I am using starship and i am trying to implement transient prompt in it.

```zsh
# Transient fns in zsh
function zle-line-init() {
  emulate -L zsh
  [[ $CONTEXT == start ]] || return 0

  while true; do
    zle .recursive-edit

    local -i ret=$?
    [[ $ret == 0 && $KEYS == $'\4' ]] || break
    [[ -o ignore_eof ]] || exit 0
  done

  local saved_prompt=$PROMPT
  local saved_rprompt=$RPROMPT
  PROMPT='ﬦ '
  RPROMPT=''
  zle .reset-prompt
  PROMPT=$saved_prompt
  RPROMPT=$saved_rprompt

  if (( ret )); then
    zle .send-break
  else
    zle .accept-line
  fi
  return ret
}

zle -N zle-line-init

```

This seems to work as expected, just one issue I am facing is that the cursor is not visible when the prompt is initialized at first.
After that, the prompt works normally
I was thinking to create a discussion on starship, but the cursor is visible if I use `eval "$( starship init zsh --print-full-init )"`.

By initialized at first I mean the cursor is not visible when the shell is sourced and if I issue a command like clear or something it works fine then.

**Comment by semi710** [comment]
Major culprit seems to be zsh-vi-mode it hijacks many fns for unknown reasons.

but 
```zsh
# Prompt
# znap eval starship 'starship init zsh --print-full-init'
# znap prompt
eval "$( starship init zsh --print-full-init )" # znap needs fix

```

if i use eval then cursor works but if i use znap prompt then it fails

**Reply by marlonrichert** [reply]
I'm not able to reproduce that:
```sh
% cd $(mktemp -d)
% git clone --depth 1 -- https://github.com/marlonrichert/zsh-snap.git
Cloning into 'zsh-snap'...
remote: Enumerating objects: 44, done.
remote: Counting objects: 100% (44/44), done.
remote: Compressing objects: 100% (40/40), done.
remote: Total 44 (delta 0), reused 16 (delta 0), pack-reused 0
Receiving objects: 100% (44/44), 20.87 KiB | 6.96 MiB/s, done.
% > .zshrc <<EOF
PS1='%# ' PS2= RPS2='%^'; setopt transientrprompt interactivecomments
source $PWD/zsh-snap/znap.zsh
znap eval starship 'starship init zsh --print-full-init'
znap prompt
EOF
% env -i HOME=$PWD PATH=$PATH TERM=$TERM ${TERMINFO:+TERMINFO=$TERMINFO} zsh -d
znap eval: generating cache for 'starship init zsh --print-full-init'

~ 
❯ _
```

It shows the cursor just fine.

**Reply by semi710** [reply]
i am sorry for the less reference 
here is my [zshrc](https://github.com/niksingh710/cli/blob/master/shell/.zshrc) transient works fine but if i use znap prompt the the cursor is not visible for the first instance.

**Reply by marlonrichert** [reply]
Please narrow it down to the smallest set of commands that can reproduce the problem. Otherwise, it's impossible to help you.

**Reply by semi710** [reply]
```zsh
function zle-line-init() {
  emulate -L zsh
  [[ $CONTEXT == start ]] || return 0

  while true; do
    zle .recursive-edit

    local -i ret=$?
    [[ $ret == 0 && $KEYS == $'\4' ]] || break
    [[ -o ignore_eof ]] || exit 0
  done

  local saved_prompt=$PROMPT
  local saved_rprompt=$RPROMPT
  PROMPT='ﬦ '
  RPROMPT=''
  zle .reset-prompt
  PROMPT=$saved_prompt
  RPROMPT=$saved_rprompt

  if (( ret )); then
    zle .send-break
  else
    zle .accept-line
  fi
  return ret
}

zle -N zle-line-init
precmd() { print "" }

plugins_path="$HOME/.zsh_plugins"

[ -d "$plugins_path" ] || mkdir -p "$plugins_path"
[[ -r "$plugins_path/znap/znap.zsh" ]] ||
  git clone --depth 1 -- \
      https://github.com/marlonrichert/zsh-snap.git "$plugins_path/znap/"

source "$plugins_path/znap/znap.zsh"

# Prompt
znap eval starship 'starship init zsh --print-full-init'
znap prompt

znap source softmoth/zsh-vim-mode
```


This is the minimal config I use and able to reproduce the bug.

When the shell starts there is no cursor but as soon as any command gets executed cursor works.

If I replace the lines 
```zsh

# Prompt
znap eval starship 'starship init zsh --print-full-init'
znap prompt

```
with 
```zsh
eval "$( starship init zsh --print-full-init )"
```
It works as expected.

**Reply by marlonrichert** [reply]
I'm planning to deprecate `znap eval`. There are more problems with it than this. I will replace it with a command that simply caches things and nothing else.

In the meantime, though, you should be fine using simply `eval` for this. Running `init zsh --print-full-init` finishes in 0.01 seconds for me. Caching that output is not going to improve the speed much.
